nifi-nar-bundles/nifi-standard-services/nifi-dbcp-service-bundle/nifi-dbcp-service/src/test/java/org/apache/nifi/dbcp/TestDBCPConnectionPoolLookup.java

```diff
@@ -0,0 +1,182 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.dbcp;
+
+import org.apache.nifi.controller.AbstractControllerService;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.reporting.InitializationException;
+import org.apache.nifi.util.TestRunner;
+import org.apache.nifi.util.TestRunners;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.sql.Connection;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TestDBCPConnectionPoolLookup {
+
+    private MockConnection connectionA;
+    private MockConnection connectionB;
+
+    private MockDBCPService dbcpServiceA;
+    private MockDBCPService dbcpServiceB;
+
+    private DBCPService dbcpLookupService;
+    private TestRunner runner;
+
+    @Before
+    public void setup() throws InitializationException {
+        connectionA = mock(MockConnection.class);
+        when(connectionA.getName()).thenReturn("A");
+
+        connectionB = mock(MockConnection.class);
+        when(connectionB.getName()).thenReturn("B");
+
+        dbcpServiceA = new MockDBCPService(connectionA);
+        dbcpServiceB = new MockDBCPService(connectionB);
+
+        dbcpLookupService = new DBCPConnectionPoolLookup();
+
+        runner = TestRunners.newTestRunner(TestProcessor.class);
+
+        final String dbcpServiceAIdentifier = "dbcp-a";
+        runner.addControllerService(dbcpServiceAIdentifier, dbcpServiceA);
+
+        final String dbcpServiceBIdentifier = "dbcp-b";
+        runner.addControllerService(dbcpServiceBIdentifier, dbcpServiceB);
+
+        runner.addControllerService("dbcp-lookup", dbcpLookupService);
+        runner.setProperty(dbcpLookupService, "a", dbcpServiceAIdentifier);
+        runner.setProperty(dbcpLookupService, "b", dbcpServiceBIdentifier);
+
+        runner.enableControllerService(dbcpServiceA);
+        runner.enableControllerService(dbcpServiceB);
+        runner.enableControllerService(dbcpLookupService);
+
+    }
+
+    @Test
+    public void testLookupServiceA() {
+        final Map<String,String> attributes = new HashMap<>();
+        attributes.put(DBCPConnectionPoolLookup.DATABASE_NAME_ATTRIBUTE, "a");
+
+        final Connection connection = dbcpLookupService.getConnection(attributes);
+        assertNotNull(connection);
+        assertTrue(connection instanceof MockConnection);
+
+        final MockConnection mockConnection = (MockConnection)connection;
+        assertEquals(connectionA.getName(), mockConnection.getName());
+    }
+
+    @Test
+    public void testLookupServiceB() {
```

Comment:
Our other "unit" tests use Derby, what about setting a service A to point to something that wouldn't connect, and B to a Derby, then make a valid query? It's more of an integration test but in this area it's not without precedent :)

Reply:
@mattyb149 Since it's a delegate, I focused on reviewing the unit tests because I felt that that would be enough here. The unit tests look correct and thorough to me.

Reply:
I felt like the main thing to test here was the logic of selecting/returning the correct connection... whether or not that connection can do something is specific to the impl it came from which I figured we already had tests for.

FWIW the live flow I tested against had DBCPConnectionPools for MySQL and Postgres, and then sent flow files with the database.name attribute into a single ExecuteSQL and verified the SQL was executed against the correct DB.